[BE] feat: 예외에 대한 로그 작성

nayonsoso opened the pull request:
<!-- 제목: [BE/FE/All] (기능 등) -->
<!-- 아래에 이슈 번호를 매겨주세요 -->

- resolves #250 

---

### 🚀 어떤 기능을 구현했나요 ?
- 예외에 대한 로그를 찍었습니다.
- 발생하는 예외 종류별로 분류해보자면,

1. 커스텀 예외
커스텀 예외 클래스에서 '로그 발생 상황'에 대한 로그를 찍습니다.
그리고 컨트롤러 어드바이스에서 로그 클래스와 스택 트레이스를 찍습니다.
로그 수준은 INFO 입니다.


```
// 🐟 이거랑
2024-08-08 15:28:52.040  INFO 6968 [nio-8080-exec-2] wGroupNotFoundByGroupAccessCodeException : ReviewGroup not found by groupAccessCode - groupAccessCode: fgasfd

// 🐟 이거 둘 다 찍힘
2024-08-08 15:28:52.041  INFO 6968 [nio-8080-exec-2] reviewme.global.GlobalExceptionHandler   : NotFoundException is occurred - ReviewGroupNotFoundByGroupAccessCodeException

reviewme.review.domain.exception.ReviewGroupNotFoundByGroupAccessCodeException: null
	at reviewme.review.service.ReviewService.lambda$findReceivedReviews$11(ReviewService.java:144)
	at java.base/java.util.Optional.orElseThrow(Optional.java:403)
	at reviewme.review.service.ReviewService.findReceivedReviews(ReviewService.java:144)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:354)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:196)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:768)
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:123)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:392)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:768)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:720)
	at reviewme.review.service.ReviewService$$SpringCGLIB$$0.findReceivedReviews(<generated>)
	at reviewme.review.controller.ReviewController.findReceivedReviews(ReviewController.java:38)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:255)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:188)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:118)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:926)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:831)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1089)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:979)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:903)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:564)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:885)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:658)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:195)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:51)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.springframework.web.filter.ServerHttpObservationFilter.doFilterInternal(ServerHttpObservationFilter.java:107)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:167)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:90)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:482)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:115)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:344)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:389)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:904)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1741)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1190)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:63)
	at java.base/java.lang.Thread.run(Thread.java:840)
```


2. 스프링 예외
컨트롤러 어드바이스에서 예외의 종류와 스프링에서 제공하는 기본 메세지를 찍습니다.
로그 수준은 INFO 입니다.

```
2024-08-08 15:28:22.595  INFO 6968 [nio-8080-exec-1] reviewme.global.GlobalExceptionHandler   : Spring error is occurred - HttpRequestMethodNotSupportedException: Request method 'PATCH' is not supported
```

3. 서버 예외

로그 수준은 ERROR 입니다.
스택 트레이스를 찍습니다.

### 🔥 어떻게 해결했나요 ?
- 스택 트레이스로 로그 발생 위치를 나타내고, 커스텀 예외 안의 로그를 통해서 예외를 발생하게 하는 변수에 대해 알려주는 완벽한 로그를 만들어냈습니다~

### 📝 어떤 부분에 집중해서 리뷰해야 할까요?
- 놓친 부분이 있는지 검토해주세요!

### 📚 참고 자료, 할 말

nayonsoso committed:
feat: Controller advice 에 로깅 추가

nayonsoso committed:
refactor: 스택 트레이스 로깅 추가

nayonsoso committed:
feat: 리뷰그룹 생성시 길이 검증 로깅 추가

nayonsoso committed:
feat: 답변 길이 검증 로깅 추가

nayonsoso committed:
feat: 리뷰 그룹 코드 검증 검증 로깅 추가

nayonsoso committed:
feat: 리뷰 조회 검증 검증 로깅 추가

nayonsoso committed:
feat: 선택된 키워드 존재하지 않는 검증 로깅 추가

nayonsoso committed:
feat: 중복 선택된 키워드 검증 로깅 추가

nayonsoso committed:
feat: 키워드 조회 검증 로깅 추가

nayonsoso committed:
feat: 선택 키워드 갯수 검증 로깅 추가

nayonsoso committed:
feat: 선택된 질문 중복 검증 로깅 추가

nayonsoso committed:
feat: 질문 조회 검증 로깅 추가

nayonsoso committed:
feat: 중복 질문 검증 로깅 추가

nayonsoso committed:
feat: 스프링 발생 예외 로깅에 메세지 추가

nayonsoso committed:
feat: 인코딩 설정

donghoony commented on backend/src/main/java/reviewme/global/GlobalExceptionHandler.java:
```diff
@@ -110,7 +110,8 @@ public ProblemDetail handleMethodArgumentNotValid(MethodArgumentNotValidExceptio
                 ex.getClass().getSuperclass().getSimpleName(),
-                ex.getClass().getSimpleName());
+                ex.getClass().getSimpleName(),
+                ex);
```
```suggestion
                ex
        );
```

donghoony commented on backend/src/main/java/reviewme/review/domain/exception/ReviewIsNotInReviewGroupException.java:
```diff
@@ -1,10 +1,13 @@
-    public ReviewIsNotInReviewGroupException() {
+    public ReviewIsNotInReviewGroupException(long reviewId, long reviewGroupId) {
         super("리뷰 그룹에 해당하는 리뷰가 아니에요.");
+        log.info("Review is not in review group - reviewId: {},reviewGroupId: {}", reviewId, reviewGroupId);
```
띄어쓰기 컨벤션 부탁해요

donghoony commented on backend/src/main/java/reviewme/keyword/domain/exception/KeywordsNotFoundException.java:
```diff
@@ -0,0 +1,14 @@
+import reviewme.global.exception.NotFoundException;
+
+@Slf4j
+public class KeywordsNotFoundException extends NotFoundException {
```
네이밍 한 번 봐주세용~

Kimprodp reviewed (approved):
(●'◡'●) 고생했슴니다 sancho

donghoony reviewed (approved):
👍🏻

nayonsoso committed:
style: 개행 수정

Co-authored-by: Donghoon Lee <aru0504@naver.com>

nayonsoso committed:
style: 개행 및 공백 수정

hyeonji1220 commented on backend/src/main/java/reviewme/global/GlobalExceptionHandler.java:
```diff
@@ -104,4 +112,20 @@ public ProblemDetail handleMethodArgumentNotValid(MethodArgumentNotValidExceptio
+    }
+
+    private void logSpringException(Exception ex) {
+        log.info("Spring error is occurred - {}: {}", ex.getClass().getSimpleName(), ex.getLocalizedMessage());
```
스프링 기본 예외라 저희가 설정해준 예외 메시지도 따로 없어서 예외에 대한 간단한 설명 메시지만 로그에 남게했군요👍

hyeonji1220 reviewed (approved):
산초 머리에서 불난다 수고했어~~

nayonsoso committed:
refactor: 불필요한 검증 제거

- 선택된 키워드와 질문이 DB에 있는지를 validator 에서 검증한 후에도, repository.getById 를 할 때 한번 더 검증이 들어간다. 따라서 'DB에 있는지'에 대한 검증을 validator 에서 할 필요는 없다는 판단 하에 해당 로직을 삭제한다.

hyeonji1220 reviewed (approved):
산초초!!! 약속했잖아!
![B948F5FE-41BD-4865-8234-EEDE55A1EF9E](https://github.com/user-attachments/assets/31187b43-ad79-4188-86cd-1c5563a9d4e6)

Kimprodp reviewed (approved):
![IDon'TKnowCryGIF](https://github.com/user-attachments/assets/82ea0582-fdf9-46e3-a262-ed2aaae35a6a)
